// testops/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestOpsMedia = exports.ExperimentRunStatus = exports.createExperiments = exports.Experiments = exports.ExperimentsClient = exports.createExperimentRuns = exports.ExperimentRuns = exports.ExperimentRunsClient = exports.createDatasetRuns = exports.DatasetRunType = exports.DatasetRunStatus = exports.DatasetRuns = exports.DatasetRunsClient = exports.Datasets = exports.DatasetsClient = void 0;
/**
 * TestOps NodeJS SDK
 * A minimal wrapper around Langfuse NodeJS SDK for TestOps API integration
 * Uses native Langfuse API with TestOps-specific enhancements
 */
const MAX_EVENT_SIZE_BYTES = 5000000; // 5MB event limit
if (!process.env.LANGFUSE_MAX_EVENT_SIZE_BYTES) {
    process.env.LANGFUSE_MAX_EVENT_SIZE_BYTES = `${MAX_EVENT_SIZE_BYTES}`;
}
const client_1 = require("./client");
__exportStar(require("./client"), exports);
var datasets_1 = require("./datasets");
Object.defineProperty(exports, "DatasetsClient", { enumerable: true, get: function () { return datasets_1.DatasetsClient; } });
Object.defineProperty(exports, "Datasets", { enumerable: true, get: function () { return datasets_1.Datasets; } });
var dataset_runs_1 = require("./dataset-runs");
Object.defineProperty(exports, "DatasetRunsClient", { enumerable: true, get: function () { return dataset_runs_1.DatasetRunsClient; } });
Object.defineProperty(exports, "DatasetRuns", { enumerable: true, get: function () { return dataset_runs_1.DatasetRuns; } });
Object.defineProperty(exports, "DatasetRunStatus", { enumerable: true, get: function () { return dataset_runs_1.DatasetRunStatus; } });
Object.defineProperty(exports, "DatasetRunType", { enumerable: true, get: function () { return dataset_runs_1.DatasetRunType; } });
Object.defineProperty(exports, "createDatasetRuns", { enumerable: true, get: function () { return dataset_runs_1.createDatasetRuns; } });
var experiment_runs_1 = require("./experiment-runs");
Object.defineProperty(exports, "ExperimentRunsClient", { enumerable: true, get: function () { return experiment_runs_1.ExperimentRunsClient; } });
Object.defineProperty(exports, "ExperimentRuns", { enumerable: true, get: function () { return experiment_runs_1.ExperimentRuns; } });
Object.defineProperty(exports, "createExperimentRuns", { enumerable: true, get: function () { return experiment_runs_1.createExperimentRuns; } });
var experiments_1 = require("./experiments");
Object.defineProperty(exports, "ExperimentsClient", { enumerable: true, get: function () { return experiments_1.ExperimentsClient; } });
Object.defineProperty(exports, "Experiments", { enumerable: true, get: function () { return experiments_1.Experiments; } });
Object.defineProperty(exports, "createExperiments", { enumerable: true, get: function () { return experiments_1.createExperiments; } });
var experiment_types_1 = require("./types/experiment-types");
Object.defineProperty(exports, "ExperimentRunStatus", { enumerable: true, get: function () { return experiment_types_1.ExperimentRunStatus; } });
const dynamicImport = (module) => {
    return Promise.resolve(`${module}`).then(s => __importStar(require(s)));
};
class TestOpsMedia extends client_1.TestOpsMediaBase {
    constructor(...args) {
        super(...args);
        dynamicImport('node:crypto')
            .then((importedCrypto) => {
            this.cryptoModule = importedCrypto;
        })
            // fallback to crypto module, which is available in browsers
            .catch(() => dynamicImport('crypto'))
            .then((importedCrypto) => {
            this.cryptoModule = importedCrypto;
        })
            .catch(() => {
            this.cryptoModule = undefined;
        });
    }
    get contentSha256Hash() {
        if (!this._contentBytes) {
            return undefined;
        }
        try {
            return this.cryptoModule
                ?.createHash('sha256')
                ?.update(this._contentBytes)
                ?.digest('base64');
        }
        catch (error) {
            return undefined;
        }
    }
}
exports.TestOpsMedia = TestOpsMedia;
